test/apps/iodemo/io_demo.cc

```diff
@@ -750,16 +769,56 @@ class DemoClient : public P2pDemoCommon {
                     << " conn " << conn;
 
         if (msg->op >= IO_COMP_MIN) {
+            assert(msg->op == IO_WRITE_COMP);
             ++_num_completed;
+            ++_server_info[get_server_index(conn)].num_completed[IO_WRITE];
+        }
+    }
+
+    long get_num_uncompleted(int server_index) const {
+        return _server_info[server_index].num_sent -
+               (_server_info[server_index].num_completed[IO_READ] +
+                _server_info[server_index].num_completed[IO_WRITE]);
+    }
+
+    static void reset_server_info(server_info_t& server_info) {
+        server_info.conn                   = NULL;
+        server_info.num_sent               = 0;
+        for (int op = 0; op < IO_OP_MAX; ++op) {
+            server_info.num_completed[op]  = 0;
+            server_info.prev_completed[op] = 0;
         }
     }
 
     virtual void dispatch_connection_error(UcxConnection *conn) {
         LOG << "setting error flag on connection " << conn;
-        _status = ERROR;
+        int server_index           = get_server_index(conn);
+        server_info_t& server_info = _server_info[server_index];
+
+        // Don't wait from completions on this connection
+        _num_sent -= get_num_uncompleted(server_index);
+
+        // Remove connection pointer
+        _conn_idx.erase(conn);
```

Comment:
why `_conn_idx` is needed? as far as I understand we can't have more 1 conn to server at the moment, so IMO better to have 1:1 link and probably some garbage conn storage

Reply:
it's lookup from ptr to index since the conn_id inside UcxConnection is: 1.private 2. incremented for every new connection to avoid matching with other connection (we can change this after moving io_demo to use active messages instead of tags)

Reply:
I understand index -> ptr lookup, but why ptr to index is needed? can assign index to conn at its start?

Reply:
- rename to server_index_lookup
- can remove conn_id from UcxConnection after moving to UCP AMs

Reply:
Renamed.